tests/std/tests/GH_001964_constexpr_generate_canonical/test.cpp

```diff
@@ -0,0 +1,57 @@
+#include <cassert>
+#include <cmath>
+#include <cstdint>
+#include <random>
+#include <utility>
+
+using namespace std;
+
+int naive_iterations(const int bits, const uint64_t gmin, const uint64_t gmax) {
+    // Naive implementation of [rand.util.canonical]. Note that for large values of range, it's possible that
+    // log2(range) == bits when range < 2^bits. This can lead to incorrect results, so we can't use this function as
+    // a reference for all values.
+
+    const double range = static_cast<double>(gmax) - static_cast<double>(gmin) + 1.0;
+    return max(1, static_cast<int>(ceil(static_cast<double>(bits) / log2(static_cast<double>(range)))));
+}
+
+void test(const int target_bits) {
+    // Increase the range until the number of iterations repeats.
+    uint64_t range = 2;
+    int k          = 0;
+    int prev_k     = -1;
+    while (k != prev_k) {
+        prev_k       = exchange(k, naive_iterations(target_bits, 1, range));
+        const int k1 = _Generate_canonical_iterations(target_bits, 1, range);
+        assert(k == k1);
+        ++range;
+    }
+
+    // Now only check the crossover points, where incrementing the range actually causes the number of iterations to
+    // increase.
+    while (--k) {

```

Comment:
Style nitpick: We conventionally avoid testing integers directly for non-zeroness. (Admittedly this is test code where we're less rigorous about conventions, but this one is pretty uniformly followed.)